csrc/runtime/fusion_kernel_runtime.cpp

```diff
@@ -445,10 +449,39 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
             std::vector<Val*>{out_clone});
         hic->pushBackTopLevelExprs(launch_kernel);
       } else {
-        // push back segment's exprs into the container as top level expressions
-        for (auto* expr : group_to_run->exprs()) {
-          auto cloned_expr = ir_cloner.clone(expr);
-          hic->pushBackTopLevelExprs(cloned_expr);
+        const bool is_resharding = std::any_of(
+            group_to_run->exprs().begin(), group_to_run->exprs().end(), [](auto expr) {
+              return isResharding(expr);
```

Comment:
isResharding is expensive (it uses simplifyExpr eg), so I'd try to avoid it whenever possible. I believe you can check group_to_run->schedulerType() instead.